rcl_action/include/rcl_action/goal_handle.h

```diff
@@ -0,0 +1,347 @@
+/**
+ * This is a non-blocking call.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[inout] goal_handle struct containing goal state to transition
+ * \param[in] goal_event the event used to transition the goal state
+ * \return `RCL_RET_OK` if the goal state was updated successfully, or
+ * \return `RCL_RET_ACTION_GOAL_EVENT_INVALID` if the goal event is invalid, or
+ * \return `RCL_RET_ACTION_GOAL_HANDLE_INVALID` if the goal handle is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_update_goal_state(
+  rcl_action_goal_handle_t * goal_handle,
+  const rcl_action_goal_event_t goal_event);
+
+/// Get the ID of a goal using a rcl_action_goal_handle_t.
+/**
+ * This is a non-blocking call.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[in] goal_handle struct containing the goal and meta
+ * \param[out] goal_info a preallocated struct where the goal info is copied
+ * \return `RCL_RET_OK` if the goal ID was accessed successfully, or
+ * \return `RCL_RET_ACTION_GOAL_HANDLE_INVALID` if the goal handle is invalid, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_goal_handle_get_info(
+  const rcl_action_goal_handle_t * goal_handle,
+  rcl_action_goal_info_t * goal_info);
+
+/// Get the request message associated with a goal using a rcl_action_goal_handle_t.
+/**
+ * This is a non-blocking call.
+ *
+ * It is the job of the caller to ensure that the type of the `ros_goal`
+ * parameter and the type associate with the goal handle, via the type
+ * support, match.
+ * Passing a different type produces undefined behavior and cannot
+ * be checked by this function and therefore no deliberate error will occur.
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Maybe [1]
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ * <i>[1] only if required when filling the goal message, avoided for fixed sizes</i>
+ *
+ * \param[in] goal_handle struct containing the goal and metadata
+ * \param[out] ros_goal a preallocated struct where the goal message is copied
+ * \return `RCL_RET_OK` if the goal ID was accessed successfully, or
+ * \return `RCL_RET_ACTION_GOAL_HANDLE_INVALID` if the goal handle is invalid, or
+ * \return `RCL_RET_BAD_ALLOC` if allocating memory failed, or
+ * \return `RCL_RET_ERROR` if an unspecified error occurs.
+ */
+RCL_PUBLIC
+RCL_WARN_UNUSED
+rcl_ret_t
+rcl_action_goal_handle_get_message(
+  const rcl_action_goal_handle_t * goal_handle,
+  void * ros_goal);
```

Comment:
Where is the original request message used later?

Reply:
I'm removing this to simplify implementation since this is the only place where typesupport for goal handles comes into play and it does not seem necessary.